File: client/src/modules/assetmanager/schemas/entity/syndicate-transaction.schemas.ts
/**
 * SyndicateTransaction Schemas
 *
 * Zod validation schemas for SyndicateTransaction model.
 * Field names and validation rules match backend exactly (snake_case).
 *
 * Backend sources:
 * - Model: /server/apps/assetmanager/models/entity_models.py
 * - Schema: /server/apps/assetmanager/schemas/entity_schemas/syndicate_transaction_schemas.py
 * - Router: /server/apps/assetmanager/subrouters/entity_subrouters/syndicate_transaction_subrouter.py
 */

import { z } from 'zod';

// ==========================================
// SyndicateTransaction Schema (Full Representation)
// ==========================================

/**
 * SyndicateTransaction schema - full representation
 * Used for GET operations (single and list)
 *
 * Backend equivalent: class SyndicateTransaction(BaseModel)
 */
export const SyndicateTransactionSchema = z.object({
  id: z.number(),
  syndicate_id: z.number(),
  // Transaction type — 'transfer' (member sells to another), 'allocation_change' (manager adjusts)
  transaction_type: z.string(),
  // Parties
  seller_entity_id: z.number(),
  buyer_entity_id: z.number(),
  // Transfer details
  ownership_percentage: z.number(),
  amount: z.number().nullable(),
  // Status flow: pending_buyer → pending_manager → completed (or rejected at any step)
  status: z.string(),
  notes: z.string().nullable(),
  // Timestamps
  requested_at: z.string(), // ISO datetime string from backend
  completed_at: z.string().nullable(),
  created_at: z.string(), // ISO datetime string from backend
  updated_at: z.string().nullable(),
});

// ==========================================
// Input Schemas
// ==========================================

/**
 * Schema for creating a new syndicate transaction (POST)
 * Excludes: id, requested_at, created_at, updated_at (auto-generated)
 *
 * Backend equivalent: class CreateSyndicateTransaction(BaseModel)
 */
export const CreateSyndicateTransactionSchema = z.object({
  syndicate_id: z.number(),
  transaction_type: z.string(),
  seller_entity_id: z.number(),
  buyer_entity_id: z.number(),
  ownership_percentage: z.number(),
  amount: z.number().nullable().optional(),
  status: z.string().default('pending_buyer').optional(),
  notes: z.string().nullable().optional(),
});

/**
 * Schema for updating a syndicate transaction (PUT/PATCH)
 * All fields optional to support partial updates
 *
 * Backend equivalent: class UpdateSyndicateTransaction(BaseModel)
 */
export const UpdateSyndicateTransactionSchema = z.object({
  transaction_type: z.string().optional(),
  seller_entity_id: z.number().optional(),
  buyer_entity_id: z.number().optional(),
  ownership_percentage: z.number().optional(),
  amount: z.number().nullable().optional(),
  status: z.string().optional(),
  notes: z.string().nullable().optional(),
  completed_at: z.string().nullable().optional(),
});

// ==========================================
// Type Exports
// ==========================================

export type SyndicateTransaction = z.infer<typeof SyndicateTransactionSchema>;
export type CreateSyndicateTransaction = z.infer<typeof CreateSyndicateTransactionSchema>;
export type UpdateSyndicateTransaction = z.infer<typeof UpdateSyndicateTransactionSchema>;

// ==========================================
// Response Types
// ==========================================

/**
 * Response containing a single syndicate transaction
 * Backend equivalent: class SyndicateTransactionResponse(BaseModel)
 */
export type SyndicateTransactionResponse = {
  success: boolean;
  data?: SyndicateTransaction;
  error?: string;
};

/**
 * Response containing multiple syndicate transactions
 * Backend equivalent: class SyndicateTransactionsResponse(BaseModel)
 */
export type SyndicateTransactionsResponse = {
  success: boolean;
  data?: SyndicateTransaction[];
  error?: string;
};

File: server/apps/assetmanager/schemas/entity_schemas/syndicate_transaction_schemas.py
"""
SyndicateTransaction Schemas

Pydantic schemas for the SyndicateTransaction model following simplified schema guidelines.
"""

from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime

# ==========================================
# SyndicateTransaction Schema (Full Representation)
# ==========================================

class SyndicateTransaction(BaseModel):
    """SyndicateTransaction schema - full representation"""
    id: int
    syndicate_id: int = Field(description="Syndicate ID")
    transaction_type: str = Field(description="Transaction type (transfer, allocation_change)")
    seller_entity_id: int = Field(description="Seller entity ID")
    buyer_entity_id: int = Field(description="Buyer entity ID")
    ownership_percentage: float = Field(description="Ownership percentage being transferred")
    amount: float | None = Field(None, description="Transaction amount (price)")
    status: str = Field(description="Transaction status (pending_buyer, pending_manager, completed, rejected)")
    notes: str | None = Field(None, description="Transaction notes")
    requested_at: datetime
    completed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)

# ==========================================
# Input Schemas
# ==========================================

class CreateSyndicateTransaction(BaseModel):
    """Schema for creating a new syndicate transaction"""
    syndicate_id: int = Field(description="Syndicate ID")
    transaction_type: str = Field(description="Transaction type (transfer, allocation_change)")
    seller_entity_id: int = Field(description="Seller entity ID")
    buyer_entity_id: int = Field(description="Buyer entity ID")
    ownership_percentage: float = Field(description="Ownership percentage being transferred")
    amount: float | None = Field(None, description="Transaction amount (price)")
    status: str = Field(default="pending_buyer", description="Transaction status")
    notes: str | None = Field(None, description="Transaction notes")

class UpdateSyndicateTransaction(BaseModel):
    """Schema for updating a syndicate transaction"""
    transaction_type: str | None = None
    seller_entity_id: int | None = None
    buyer_entity_id: int | None = None
    ownership_percentage: float | None = None
    amount: float | None = None
    status: str | None = None
    notes: str | None = None
    completed_at: datetime | None = None

# ==========================================
# Response Types
# ==========================================

class SyndicateTransactionResponse(BaseModel):
    """Response containing a single syndicate transaction"""
    success: bool
    data: SyndicateTransaction | None = None
    error: str | None = None

class SyndicateTransactionsResponse(BaseModel):
    """Response containing multiple syndicate transactions"""
    success: bool
    data: list[SyndicateTransaction] | None = None
    error: str | None = None
